File: packages/front/src/lib/contracts/dto/market.ts
import type {
  AmbiguityLevel,
  EventStatus,
  MarketStatus,
  ResourceVersion,
  TradabilityStatus,
} from "./primitives";

export type MarketDto = ResourceVersion & {
  question: string;
  category: string;
  slug?: string | null;
  status: MarketStatus;
  best_bid: string;
  best_ask: string;
  mid_price: string;
  volume_24h: string;
  ambiguity_level: AmbiguityLevel;
  tradability_status: TradabilityStatus;
  resolution_source: string;
  edge_case_notes: string[];
  polymarket_condition_id?: string | null;
  polymarket_yes_asset_id?: string | null;
  polymarket_no_asset_id?: string | null;
  updated_at: string;
};

export type EventDto = ResourceVersion & {
  source: string;
  summary: string;
  relevance_score: string;
  confidence: string;
  status: EventStatus;
  related_market_ids: string[];
  reason_trace: string;
  created_at: string;
  updated_at: string;
};

File: packages/backend/crates/contracts/src/dto/market.rs
// Market and event DTOs, including the market list envelope.

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketListResponse {
    pub data: Vec<MarketData>,
    pub total_count: i64,
    pub meta: ApiMeta,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketCategoryData {
    pub id: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketData {
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub slug: Option<String>,
    pub question: String,
    pub category: String,
    pub status: MarketStatus,
    pub best_bid: Probability,
    pub best_ask: Probability,
    pub mid_price: Probability,
    pub volume_24h: UsdAmount,
    pub liquidity_usd: UsdAmount,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[serde(with = "time::serde::rfc3339::option")]
    pub end_at: Option<OffsetDateTime>,
    pub ambiguity_level: AmbiguityLevel,
    pub tradability_status: TradabilityStatus,
    pub resolution_source: String,
    pub edge_case_notes: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub polymarket_condition_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub polymarket_yes_asset_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub polymarket_no_asset_id: Option<String>,
    #[serde(with = "time::serde::rfc3339")]
    pub updated_at: OffsetDateTime,
    pub version: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EventData {
    pub id: String,
    pub source: String,
    pub summary: String,
    pub relevance_score: Probability,
    pub confidence: Probability,
    pub status: EventStatus,
    pub related_market_ids: Vec<String>,
    pub reason_trace: String,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub updated_at: OffsetDateTime,
    pub version: i64,
}
